checkout-core/src/main/java/com/adyen/checkout/core/util/FileDownloader.kt

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2022 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by atef on 16/11/2022.
+ */
+
+package com.adyen.checkout.core.util
+
+import android.Manifest
+import android.content.ContentValues
+import android.content.Context
+import android.os.Build
+import android.provider.MediaStore
+import androidx.annotation.RequiresApi
+import androidx.annotation.RequiresPermission
+import androidx.annotation.RestrictTo
+import com.adyen.checkout.core.exception.CheckoutException
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.FileOutputStream
+import java.net.MalformedURLException
+import java.net.URL
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+@Suppress("LongParameterList")
+class FileDownloader {
+
+    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    suspend fun download(
+        context: Context,
+        stringUrl: String,
+        fileName: String,
+        filePath: String,
+        mimeType: String? = null
+    ): Result<Unit> {
+        val url = stringUrl.toULR() ?: return Result.failure(CheckoutException("Malformed URL"))
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
```

Comment:
These two can actually be combined into one method making the class less verbose
```Kotlin
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class FileDownloader {

    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    suspend fun download(
        context: Context,
        stringUrl: String,
        fileName: String,
        filePath: String,
        mimeType: String? = null
    ): Result<Unit> = withContext(Dispatchers.IO) {
        val resolver = context.contentResolver

        val values = ContentValues().apply {
            put(MediaStore.MediaColumns.DISPLAY_NAME, fileName)
            put(MediaStore.MediaColumns.MIME_TYPE, mimeType)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                put(MediaStore.MediaColumns.RELATIVE_PATH, filePath)
            } else {
                put(MediaStore.MediaColumns.DATA, "${Environment.getExternalStorageDirectory()}/$filePath/$fileName")
            }
        }

        val target = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            MediaStore.Downloads.EXTERNAL_CONTENT_URI
        } else {
            MediaStore.Files.getContentUri("external")
        }

        val uri = resolver.insert(target, values) ?: return@withContext Result.failure(CheckoutException("URI is null"))

        val url = stringUrl.toURL() ?: return@withContext Result.failure(CheckoutException("Malformed URL"))

        url.openStream().use { input ->
            resolver.openOutputStream(uri).use { output ->
                if (output == null) return@withContext Result.failure(CheckoutException("out is null"))
                input.copyTo(output, DEFAULT_BUFFER_SIZE)
                return@withContext Result.success(Unit)
            }
        }
    }

    private fun String.toURL(): URL? {
        return try {
            URL(this)
        } catch (e: MalformedURLException) {
            Logger.e(TAG, "toURL: $e")
            null
        }
    }

    companion object {
        private val TAG = LogUtil.getTag()
    }
}
```

Reply:
let me check again as when I tried something similar to this it wasn't working on below api 28.